### PR TITLE
[DONE] Companion mission main loop, struct instead of string

### PR DESCRIPTION
**Pull request by acidia**
Part of Coolthulhu's suggestions in #24040, the main loop of the companion mission function now uses

> struct mission_entry {
      std::string id;
      std::string name_display;
      std::string dir;
      bool priority;
      bool possible;
};

To store most of the information instead of a string.  This shouldn't change anything visible to the player, decreases the number of string.find() calls, and allows translators to change the mission titles since name_display is separate from the id that is checked during comparisons.  Doesn't change anything for crafting missions since those are generated from the maps at the moment but should be moved to a new recipe group json similar to a monster group, so that modders can change what recipes are available at the kitchen, blacksmith, etc.

### CODE REVIEW

File: src/mission_companion.cpp [hunk 1]
```diff
@@ -379,14 +388,14 @@ bool talk_function::outpost_missions( npc &p, const std::string &id, const std::
...
     }
 
     if( id == "FACTION_CAMP" && om_min_level("faction_base_camp_1", om_cur ) ){
         col_missions["Gather Materials"] = om_gathering_description( p, bldg);
         bool avail = ( companion_list( p, "_faction_camp_gathering" ).size() < 3 );
-        mission_key_push( key_vectors, "Gather Materials", "", false, avail );
+        mission_key_push( mission_key_vectors, "Gather Materials", _("Gather Materials"), "", false, avail );
 
 
         col_missions["Distribute Food"] = string_format( "Notes:\n"
```
Comment (cainiaowu) [inline src/mission_companion.cpp]: While we are at it, might as well translate these mission descriptions.

File: src/mission_companion.cpp [hunk 2]
```diff
@@ -400,7 +409,7 @@ bool talk_function::outpost_missions( npc &p, const std::string &id, const std::
             "> Rots in < 2 days: 60%%\n"
             "> Rots in < 5 days: 80%%\n \n"
             "Total faction food stock: %d kcal or %d day's rations", 10 * camp_food_supply(), camp_food_supply( 0, true ) );
-        mission_key_push( key_vectors, "Distribute Food" );
+        mission_key_push( mission_key_vectors, "Distribute Food", _("Distribute Food") );
 
         col_missions["Reset Sort Points"] = string_format("Notes:\n"
```
Comment (cainiaowu) [inline src/mission_companion.cpp]: Here, and many more after it as well. You should probably use search function to find and replace all these.

**Comment by cainiaowu** [comment]
Since there are now struct, please consider move them into json file and unhardcoded them.

File: src/mission_companion.cpp (hunk 2 above)
Comment (acidia) [reply src/mission_companion.cpp]: How are you requesting these be formatted?  I'm assuming sentence structure is also going to be different, which means using the string_format structure itself isn't going to work.  There isn't a way to push the entire string_format() block to json, is there?

Comment (cainiaowu) [reply src/mission_companion.cpp]: Emm, I dont know much about coding, but could you follow npcs/missiondef.json and src/missiondef.cpp ?

So, the json object is loaded by mission_type::load() so you may need to modified it to accept your data format.

**Pull request merged**